com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/ComputeNextAppsStep.java

```diff
@@ -0,0 +1,55 @@
+package com.sap.cloud.lm.sl.cf.process.steps;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+import com.sap.cloud.lm.sl.cf.client.lib.domain.CloudApplicationExtended;
+import com.sap.cloud.lm.sl.cf.core.security.serialization.SecureSerializationFacade;
+import com.sap.cloud.lm.sl.cf.process.message.Messages;
+
+@Component("computeNextAppsStep")
+@Scope(BeanDefinition.SCOPE_PROTOTYPE)
+public class ComputeNextAppsStep extends SyncActivitiStep {
+
+    protected SecureSerializationFacade secureSerializer = new SecureSerializationFacade();
+
+    @Override
+    protected StepPhase executeStep(ExecutionWrapper execution) throws Exception {
+        getStepLogger().debug(Messages.COMPUTING_NEXT_APPS_FOR_PARALLEL_DEPLOYMENT);
+        List<CloudApplicationExtended> allApplicationsToDeploy = StepsUtil.getAppsToDeploy(execution.getContext());
+        List<CloudApplicationExtended> completedApplications = StepsUtil.getCompletedAppsInParallel(execution.getContext());
+
+        // Set next iteration data
+        List<CloudApplicationExtended> appsForNextIteration = getApplicationsForNextIteration(allApplicationsToDeploy,
+            completedApplications.stream()
+                .map(CloudApplicationExtended::getModuleName)
+                .collect(Collectors.toList()));
+        getStepLogger().debug(Messages.APPS_FOR_NEXT_PARALLEL_ITERATION, secureSerializer.toJson(appsForNextIteration));
+        StepsUtil.setAppsToDeployInParallel(execution.getContext(), appsForNextIteration);
+
+        // Mark next iteration data as completed
+        completedApplications.addAll(appsForNextIteration);
+        StepsUtil.setCompletedAppsInParallel(execution.getContext(), completedApplications);
+
+        boolean isLastIteration = allApplicationsToDeploy.stream()
+            .noneMatch(app -> completedApplications.stream()
+                .noneMatch(completedApp -> completedApp.getModuleName()
+                    .equals(app.getModuleName())));
+        StepsUtil.setIsLastParallelIteration(execution.getContext(), isLastIteration);
+        getStepLogger().debug(Messages.COMPUTED_NEXT_APPS_FOR_PARALLEL_DEPLOYMENT);
+        return StepPhase.DONE;
+    }
+
+    private List<CloudApplicationExtended> getApplicationsForNextIteration(List<CloudApplicationExtended> allApplicationsToDeploy,
+        List<String> completedModules) {
+        return allApplicationsToDeploy.stream()
+            .filter(app -> !completedModules.contains(app.getModuleName()) && (app.getDeployedAfter()
```

Comment:
Could first, the completed to be removed from the list of applications?
Then, extract all applications which have deployedAfter which contains one of the completedModules?
Isn't it be simpler that way?